### PR TITLE
8295948: Support for Zicbop/prefetch instructions on RISC-V

### PR DESCRIPTION
**Pull request by luhenry**
The OpenJDK supports generating prefetch instructions on most platforms. RISC-V supports through the Zicbop extension the use of prefetch instructions. We want to make sure we use these instructions whenever they are available.

It passes `hotspot:tier1` test suite

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295948](https://bugs.openjdk.org/browse/JDK-8295948): Support for Zicbop/prefetch instructions on RISC-V

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author) ⚠️ Review applies to [0e92909e](https://git.openjdk.org/jdk/pull/10884/files/0e92909e0d21f18a8de0fb893742d31f18a951cf)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10884/head:pull/10884` \
`$ git checkout pull/10884`

Update a local copy of the PR: \
`$ git checkout pull/10884` \
`$ git pull https://git.openjdk.org/jdk pull/10884/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10884`

View PR using the GUI difftool: \
`$ git pr show -t 10884`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10884.diff">https://git.openjdk.org/jdk/pull/10884.diff</a>

</details>

### CODE REVIEW

**Comment by RealFYang** [comment]
@luhenry : Sorry for late reply. I am looking this this now.

**Comment by luhenry** [comment]
@realfyang perfect, thank you

**Comment by luhenry** [comment]
@realfyang perfect, thank you

File: src/hotspot/cpu/riscv/riscv.ad [hunk 1]
```diff
@@ -5172,6 +5182,25 @@ instruct storeNKlass(iRegN src, memory mem)
...
+
+instruct prefetchalloc( memory mem ) %{
+  match(PrefetchAllocation mem);
+
+  ins_cost(ALU_COST * 3);
+  format %{ "prefetch_w $mem\t# Prefetch for write" %}
+
+  ins_encode %{
+    __ addi(t0, as_Register($mem$$base), $mem$$disp);
+    __ andi(t0, t0, ~(CacheLineSize - 1));
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv.ad:5197]: Do we really need to align to CacheLineSize here and in generate_prefetch?
I didn't see this restriction from the official CMO specification [1] about this requirement.
Could you please confirm that? Thanks.

[1] https://github.com/riscv/riscv-CMOs/blob/master/specifications/cmobase-v1.0.1.pdf

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5197]: It's indeed not a requirement, let me remove that `andi`.

File: src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp [hunk 2]
```diff
@@ -27,12 +27,22 @@
...
 #include "runtime/prefetch.hpp"
+#include "runtime/stubRoutines.hpp"
 
-
-inline void Prefetch::read (const void *loc, intx interval) {
+inline void Prefetch::read(const void *loc, intx interval) {
+    void (*stub)(const void*, intptr_t) =
+        (void (*)(const void*, intptr_t))StubRoutines::riscv::prefetch_r();
+    if (interval >= 0 && stub != NULL) {
+        stub(loc, interval);
```
Comment (RealFYang) [inline src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp:36]: I am not sure if it really worth it to call a stub for read / write here. It looks to me not a big issue for the case the stub tries to catch and resolve. And I see aarch64 simply plant a 'prfm' instruction for prefetching [1]. I guess we might can do the same?

File: src/hotspot/cpu/riscv/riscv.ad [hunk 3]
```diff
@@ -5172,6 +5182,24 @@ instruct storeNKlass(iRegN src, memory mem)
...
+// Must be safe to execute with invalid address (cannot fault).
+
+instruct prefetchalloc( memory mem ) %{
+  match(PrefetchAllocation mem);
+
+  ins_cost(ALU_COST * 3);
+  format %{ "prefetch_w $mem\t# Prefetch for write" %}
+
+  ins_encode %{
+    __ addi(t0, as_Register($mem$$base), $mem$$disp);
```
[the remark this answers is not in the document]
Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5196]: The offset needs to be aligned on 32 bytes (the lower 5 bits must be zero). There is then no guarantee that `$mem$$base + ($mem$$disp & ~((1<<5)-1)` is still on the same cache line. It's then easier to do a prefetch of `base+disp` with `offset = 0`.

File: src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp (hunk 2 above)
Comment (luhenry) [reply src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp:36]: We would need to check for `UseZicbop` in any case; the access to a global variable is then required.

It would be the same issue as https://github.com/openjdk/jdk/pull/10884/files/e968f7164124dcf560807c9ff7765e6f82b64cdd#diff-e3c18b8b83898e82b5a3069319df6a47468e91cc2527bf065e704a685a20f26bR5196 without the stub.

I've to admit that the `interval` naming here is confusing since no implementation ever uses it as an interval but alway as an offset. Also, the callers assume it to be an offset, like `ContiguousSpace::prepare_for_compaction` for example.

If we are to simply ignore the interval parameter, then we can also remove all uses of `PrefetchCopyIntervalInBytes`, `PrefetchScanIntervalInBytes`,

File: src/hotspot/cpu/riscv/riscv.ad (hunk 3 above)
[the remark this answers is not in the document]
Comment (RealFYang) [reply src/hotspot/cpu/riscv/riscv.ad:5196]: But what if we are passed some $mem$$disp which is multiple of 32 and thus satisfies the constraint? Then this "addi" instruction could be eliminated, right?

File: src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp (hunk 2 above)
Comment (RealFYang) [reply src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp:36]: Yes, I agree that a check for UseZicbop option would be necessary. But I still don't understand why we should implement this through a stub here. It looks to me that CPP code with inline assembly would also do. At least this could help eliminate the prologue & epilogue cost of calling the stub.

Comment (luhenry) [reply src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp:36]: We can definitely go with inline assembly here, just not sure that `prefetch.r` or `prefetch.w` would be recognized by the assembler. However, given that `prefetch.*` is an `ori` under the hood, we can simply use that.

File: src/hotspot/cpu/riscv/riscv.ad (hunk 3 above)
[the remark this answers is not in the document]
Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5196]: We want to guarantee that `$mem$$base + $mem$$disp` is on the same cache line as `$mem$$base + ($mem$$disp & ~0x1f)` or even `$mem$$base + (($mem$$disp & ~0x1f) + 0x1f)`. It's easy to find cases that trip these two possible solutions:
1. for `$mem$$base + ($mem$$disp & ~0x1f)`, if `$mem$$base = 0x30` and `$mem$$disp = 0x10`, then `$mem$$base + $mem$$disp = 0x40`, while `$mem$$base + ($mem$$disp & 0x1f) = 0x30` which are not on the same 64 bytes cache line.
2. for `$mem$$base + ($mem$$disp & ~0x1f) + 0x20`, if `$mem$$base = 0x30` and `$mem$$disp = 0x0`, then `$mem$$base + ($mem$$disp & ~0x1f) + 0x20 = 0x50` which again are not on the same 64 bytes cache line.

The simplest and most accurate solution is the `__ addi` just before. Given that it's an immediate add and the disp values are at a maximum of `64 * CacheLineSize`, it's always going to fit in `addi` immediate.

[the remark this answers is not in the document]
Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad]: I've added a check for the value of `$mem$$disp`.

File: src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp [hunk 4]
```diff
@@ -28,11 +28,18 @@
 
 #include "runtime/prefetch.hpp"
 
-
-inline void Prefetch::read (const void *loc, intx interval) {
+inline void Prefetch::read(const void *loc, intx interval) {
+    if (interval >= 0 && UseZicbop) {
```
Comment (RealFYang) [inline src/hotspot/os_cpu/linux_riscv/prefetch_linux_riscv.inline.hpp:32]: You might want to fix indentation for Prefetch::read/write here.

**Review by RealFYang — APPROVED** [review]
Updated changes looks good except for the indentation issue.

File: src/hotspot/cpu/riscv/riscv.ad [hunk 5]
```diff
@@ -5172,6 +5182,28 @@ instruct storeNKlass(iRegN src, memory mem)
   ins_pipe(istore_reg_mem);
 %}
 
+// ============================================================================
+// Prefetch instructions
+// Must be safe to execute with invalid address (cannot fault).
+
+instruct prefetchalloc( memory mem ) %{
+  match(PrefetchAllocation mem);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv.ad:5190]: PS: Should we also put this under control of option UseZicbop like you do in Prefetch::read/write? Did you checked whether those (prefetchalloc and Prefetch::read/write) will ever be used when AllocatePrefetchStyle is 0 (which is the case when UseZicbop is false).

Comment (yadongw) [reply src/hotspot/cpu/riscv/riscv.ad:5190]: The generation of PrefetchAllocationNode is controlled by AllocatePrefetchStyle. It should be fine if UseZicbop is associated with this option.

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5191]: Added `predicate(UseZicbop)` just in case.

**Review by RealFYang — APPROVED** [review]
Still looks good. Thanks.

File: src/hotspot/cpu/riscv/riscv.ad [hunk 6]
```diff
@@ -5172,6 +5182,29 @@ instruct storeNKlass(iRegN src, memory mem)
...
+  ins_cost(ALU_COST * 3);
+  format %{ "prefetch_w $mem\t# Prefetch for write" %}
+
+  ins_encode %{
+    if ((($mem$$disp) & 0x1f) == 0) {
+      __ prefetch_w(as_Register($mem$$base), $mem$$disp);
+    } else {
+      __ addi(t0, as_Register($mem$$base), $mem$$disp);
+      __ prefetch_w(t0, 0);
+    }
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/riscv.ad:5202]: It'd better to generate the case that handles the imm not fit for instructions like prefetch.w or addi.

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5202]: I'm not sure I understand what you're suggesting.

Comment (yadongw) [reply src/hotspot/cpu/riscv/riscv.ad:5202]: I mean we should handle the situation that disp beyonds 12-bit range in addi or prefetch.w.

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5202]: How about we add an assert for now? That `disp` value is driven by `AllocatePrefetchDistance`, `AllocatePrefetchStepSize`, and `AllocatePrefetchLines`/`AllocateInstancePrefetchLines`. Given the default value of these config variables, the maximum `disp` would be `3*CacheLineSize + max(3,1) * CacheLineSize` which is (on most platforms) 384 bytes. If you max-out the number of lines to prefetch, you'd get `3*CacheLineSize + 64 * CacheLineSize` or 4288 bytes which indeed requires more than 12 bit of storage.

Either we can safeguard against that in `vm_version_riscv.cpp` (given these values are quite nonsensical in themselves, given the cost/benefit of doing so many prefetches or so far). Either we can complexify the code generation here in that case.

Comment (yadongw) [reply src/hotspot/cpu/riscv/riscv.ad:5202]: I'm inclined to support the second option.

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad:5206]: Fixed.

**Comment by luhenry** [comment]
@RealFYang @yadongw could you please review+sponsor if it looks all good to you? Thanks!

File: src/hotspot/cpu/riscv/riscv.ad [hunk 7]
```diff
@@ -5172,6 +5182,33 @@ instruct storeNKlass(iRegN src, memory mem)
...
+
+instruct prefetchalloc( memory mem ) %{
+  predicate(UseZicbop);
+  match(PrefetchAllocation mem);
+
+  ins_cost(ALU_COST * 3);
+  format %{ "prefetch_w $mem\t# Prefetch for write" %}
+
+  ins_encode %{
+    if ((($mem$$disp) & 0x1f) == 0) {
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/riscv.ad:5197]: Does this branch only check the alignment, not the allowed range?

Comment (luhenry) [reply src/hotspot/cpu/riscv/riscv.ad]: Fixed.

**Review by yadongw — APPROVED** [review]
lgtm

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 8]
```diff
@@ -1917,6 +1917,10 @@ void MacroAssembler::ror_imm(Register dst, Register src, uint32_t shift, Registe
 void MacroAssembler::andi(Register Rd, Register Rn, int64_t imm, Register tmp) {
   if (is_imm_in_range(imm, 12, 0)) {
     and_imm12(Rd, Rn, imm);
+  } else if (is_power_of_2(~(imm - 1))) {
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1920]: Since this has been reworked, is this change still necessary then? I assume this won't make a difference in number of instructions emitted as compared with the final else branch?

**Review by RealFYang — APPROVED** [review]
